src/ts_emit.iced

```diff
@@ -18,8 +22,99 @@ exports.TypescriptEmitter = class TypescriptEmitter extends BaseEmitter
     @output " */"
     @output ""
 
+  output_doc : (doc) ->
+    if doc?
+      @output "/**"
+      for line in doc.split /\n/
+        @output " * " + line.replace /^\s$/, ''
+      @output " */"
 
-  # TODO: This should be removed in favor of the complete definition in base_emitter once more methods are defined
-  run : ({infiles, outfile, json, options}) ->
-    @emit_preface infiles, json, options
-    @_code
+  convert_primitive_type : (type) ->
+    map =
+      bool : "boolean"
+      bytes : "Buffer"
+      long : "number"
+      float : "number"
+      double : "number"
+      uint : "number"
+      int : "number"
+      uint64 : "number"
+      int64 : "number"
+    map[type] or type
+
+  make_map_type : (type) ->
+    "{[key: string]: #{@emit_field_type(type.values).type}}"
+
+  emit_field_type : (t, {optionalkey} = {}) ->
+    optional = optionalkey?
+    type = if typeof(t) is 'string' then @convert_primitive_type(t)
+    else if typeof(t) is 'object'
+      if Array.isArray(t)
+        if not t[0]?
+          optional = true
+          @emit_field_type(t[1]).type
+        else
+          "ERROR"
+      else if t.type is "array"
+        @emit_field_type(t.items).type + "[]"
+      else if t.type is "map"
+        @make_map_type t
+      else "ERROR"
+    else "ERROR"
```

Comment:
So it emits ERROR and that's definitely going to be a semantic error in ts? Cool.

Reply:
hmm, I should probably make these actual `Error`s, this was a copy/paste from some older go code that I think was updated to Errors.